File: 2024/16/sol.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"math"
	"os"
)

type Point struct {
	x, y int
}

type Direction struct {
	dx, dy int
}

type State struct {
	pos Point
	dir Direction
}

var TURN_CW = map[Direction]Direction{
	{1, 0}:  {0, 1},  // E -> S
	{0, 1}:  {-1, 0}, // S -> W
	{-1, 0}: {0, -1}, // W -> N
	{0, -1}: {1, 0},  // N -> E
}

var TURN_CCW = map[Direction]Direction{
	{1, 0}:  {0, -1}, // E -> N
	{0, 1}:  {1, 0},  // S -> E
	{-1, 0}: {0, 1},  // W -> S
	{0, -1}: {-1, 0}, // N -> W
}

func (r *State) forward() State {
	return State{pos: Point{r.pos.x + r.dir.dx, r.pos.y + r.dir.dy}, dir: r.dir}
}
func (r *State) back() State {
	return State{pos: Point{r.pos.x - r.dir.dx, r.pos.y - r.dir.dy}, dir: r.dir}
}
func (r *State) turn() [2]State {
	return [2]State{
		{pos: r.pos, dir: TURN_CW[r.dir]},
		{pos: r.pos, dir: TURN_CCW[r.dir]},
	}
}

type Grid struct {
	data  map[Point]bool
	w, h  int
	start Point
	end   Point

	cost     map[State]int
	min_cost int
}

func NewGrid(lines []string) *Grid {
	g := &Grid{}
	g.h = len(lines)
	g.data = make(map[Point]bool)
	if g.h > 0 {
		g.w = len(lines[0])
	}

	// Fill data map
	for y, row := range lines {
		for x, ch := range row {
			if ch == '#' {
				continue
			}
			p := Point{x, y}
			g.data[p] = true
			switch ch {
			case 'S':
				g.start = p
			case 'E':
				g.end = p
			}
		}
	}

	g.cost = make(map[State]int)
	g.min_cost = math.MaxInt32

	return g
}

func (g *Grid) at(p Point) bool {
	val, ok := g.data[p]
	return val && ok
}

// Item for priority queue
type Item struct {
	cost  int
	state State
	index int
}

// PriorityQueue implements heap.Interface
type PriorityQueue []*Item

func (pq PriorityQueue) Len() int { return len(pq) }
func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].cost < pq[j].cost
}
func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}
func (pq *PriorityQueue) Push(x interface{}) {
	item := x.(*Item)
	item.index = len(*pq)
	*pq = append(*pq, item)
}
func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	*pq = old[0 : n-1]
	return item
}

func (g *Grid) solve() bool {
	startDir := Direction{1, 0} // Facing east first
	startState := State{pos: g.start, dir: startDir}
	g.cost[startState] = 0

	pq := &PriorityQueue{}
	heap.Init(pq)
	heap.Push(pq, &Item{cost: 0, state: State{pos: g.start, dir: startDir}})

	visited := make(map[State]bool)

	for pq.Len() > 0 {
		item := heap.Pop(pq).(*Item)
		cost, state := item.cost, item.state

		if state.pos == g.end {
			g.min_cost = cost
			return true
		}

		if visited[state] {
			continue
		}
		visited[state] = true

		// Prune if we have already found a better path
		if x, ok := g.cost[state]; ok && x < cost {
			continue
		}

		// Turn
		for _, nextState := range state.turn() {
			if x, ok := g.cost[nextState]; !ok || cost+1000 < x {
				g.cost[nextState] = cost + 1000
				heap.Push(pq, &Item{cost: cost + 1000, state: nextState})
			}
		}

		// Move
		nextState := state.forward()
		if g.at(nextState.pos) {
			if x, ok := g.cost[nextState]; !ok || cost+1 < x {
				g.cost[nextState] = cost + 1
				heap.Push(pq, &Item{cost: cost + 1, state: nextState})
			}
		}
	}

	return false
}

func (g *Grid) count_best() int {
	queue := []State{}
	visited := make(map[State]bool)
	for _, d := range TURN_CCW {
		st := State{pos: g.end, dir: d}
		if x, ok := g.cost[st]; ok && x == g.min_cost {
			queue = append(queue, st)
			visited[st] = true
		}
	}

	cells := make(map[Point]bool)
	for len(queue) > 0 {
		state := queue[0]
		queue = queue[1:]
		dist := g.cost[state]
		cells[state.pos] = true

		// Backtracking turns:
		for _, prevState := range state.turn() {
			if !visited[prevState] {
				if x, ok := g.cost[prevState]; ok && x+1000 == dist {
					visited[prevState] = true
					queue = append(queue, prevState)
				}
			}
		}

		// Backtracking moves:
		prevState := state.back()
		if g.at(prevState.pos) {
			if !visited[prevState] {
				if x, ok := g.cost[prevState]; ok && x+1 == dist {
					visited[prevState] = true
					queue = append(queue, prevState)
				}
			}
		}
	}

	return len(cells)
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var data []string
	for scanner.Scan() {
		line := scanner.Text()
		data = append(data, line)
	}

	grid := NewGrid(data)
	if grid.solve() {
		fmt.Println(grid.min_cost)
		fmt.Println(grid.count_best())
	}
}

File: 2024/16/sol.py
from heapq import heappop, heappush
import dataclasses
import collections
import sys
from typing import List, Tuple, Set, Dict


@dataclasses.dataclass(frozen=True, eq=True, order=True)
class Point:
    x: int
    y: int


@dataclasses.dataclass(frozen=True, eq=True, order=True)
class Direction:
    dx: int
    dy: int


TURN_CW = {
    Direction(1, 0): Direction(0, 1),  # E -> S
    Direction(0, 1): Direction(-1, 0),  # S -> W
    Direction(-1, 0): Direction(0, -1),  # W -> N
    Direction(0, -1): Direction(1, 0),  # N -> E
}

TURN_CCW = {
    Direction(1, 0): Direction(0, -1),  # E -> N
    Direction(0, 1): Direction(1, 0),  # S -> E
    Direction(-1, 0): Direction(0, 1),  # W -> S
    Direction(0, -1): Direction(-1, 0),  # N -> W
}


@dataclasses.dataclass(frozen=True, eq=True, order=True)
class State:
    pos: Point
    dir: Direction

    def forward(self: "State") -> "State":
        return State(
            Point(self.pos.x + self.dir.dx, self.pos.y + self.dir.dy),
            self.dir,
        )

    def back(self: "State") -> "State":
        return State(
            Point(self.pos.x - self.dir.dx, self.pos.y - self.dir.dy),
            self.dir,
        )

    def turn(self: "State") -> List["State"]:
        return [State(self.pos, nd) for nd in (TURN_CW[self.dir], TURN_CCW[self.dir])]


class Grid:
    def __init__(self, lines):
        self.w, self.h = len(lines[0]), len(lines)
        self.start = None
        self.end = None
        self.data: Set[Point] = set()
        for y, row in enumerate(lines):
            for x, ch in enumerate(row):
                if ch == "#":
                    continue
                p = Point(x, y)
                self.data.add(p)
                if ch == "S":
                    self.start = p
                elif ch == "E":
                    self.end = p
        self.cost: Dict[State, int] = {}
        self.min_cost = None

    def solve(self) -> bool:
        start_dir = Direction(1, 0)  # Facing east first
        start_state = State(self.start, start_dir)
        self.cost[start_state] = 0

        pq: List[Tuple[int, State]] = []
        heappush(pq, (0, start_state))
        visited = set()
        while pq:
            cost, state = heappop(pq)
            if state.pos == self.end:
                self.min_cost = cost
                return True

            if state in visited:
                continue
            visited.add(state)

            # Prune if we have already found a better path
            if state in self.cost and self.cost[state] < cost:
                continue

            # Turn
            for next_state in state.turn():
                if next_state not in self.cost or cost + 1000 < self.cost[next_state]:
                    self.cost[next_state] = cost + 1000
                    heappush(pq, (cost + 1000, next_state))

            # Move
            next_state = state.forward()
            if next_state.pos in self.data:
                if next_state not in self.cost or cost + 1 < self.cost[next_state]:
                    self.cost[next_state] = cost + 1
                    heappush(pq, (cost + 1, next_state))
        return False

    def count_best(self):
        queue: collections.deque[State] = collections.deque()
        visited = set()
        for d in TURN_CCW:
            start_state = State(self.end, d)
            if start_state in self.cost and self.cost[start_state] == self.min_cost:
                queue.append(start_state)
                visited.add(start_state)

        cells = set()
        while queue:
            state = queue.popleft()
            dist = self.cost[state]
            cells.add(state.pos)

            # Backtracking turns:
            for prev_state in state.turn():
                if (
                    prev_state not in visited
                    and prev_state in self.cost
                    and self.cost[prev_state] + 1000 == dist
                ):
                    visited.add(prev_state)
                    queue.append(prev_state)

            # Backtracking moves:
            prev_state = state.back()
            if prev_state.pos in self.data:
                if (
                    prev_state not in visited
                    and prev_state in self.cost
                    and self.cost[prev_state] + 1 == dist
                ):
                    visited.add(prev_state)
                    queue.append(prev_state)

        return len(cells)


if __name__ == "__main__":
    lines = []

    for line in sys.stdin:
        lines.append(line.strip())

    grid = Grid(lines)
    if grid.solve():
        print(grid.min_cost)
        print(grid.count_best())
